File: connect4_ai.c
#include <stdio.h>
#include <math.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define boardSize 168
#define size 4

int isValid(int (*)[6][7], int);
void makeMove(int (*)[6][7], int, int);
int isTerminal(int (*)[6][7]);
int evaluateSegment(int, int, int, int);
int evaluatePosition(int (*)[6][7]);
struct evaluation minimax(int (*)[6][7], int, int, int, int);
void printBoard(int (*)[6][7]);

const int AI = 1;
const int OPPONENT = -1;
int board[6][7] =  {{0, 0, 0, 0, 0, 0, 0},
			    	{0, 0, 0, 0, 0, 0, 0},
			    	{0, 0, 0, 0, 0, 0, 0},
				    {0, 0, 0, 0, 0, 0, 0},
			    	{0, 0, 0, 0, 0, 0, 0},
			    	{0, 0, 0, 0, 0, 0, 0}};

struct evaluation{
	int eval;
	int bestMove;
};


void main(){
	/*struct evaluation score = minimax(board, AI, 5, -INFINITY, INFINITY);
	printf("%i\n", score.eval);
	printf("%i\n", score.bestMove);*/
	/*struct evaluation e = minimax(&board, AI, 1, -INFINITY, INFINITY);
	printf("%i\n", e.eval);
	printf("%i\n", e.bestMove);*/
	int whoFirst = 0;
	do {
		printf("Who Goes first?:\n[1] Human\n[2] AI\n>>");
		scanf("%i", &whoFirst);
	}
	while(whoFirst == 0);
	
	if (whoFirst == 2) {
		makeMove(&board, 4-1, AI);
	}
	
	int gameOver = 0;
	while(gameOver == 0){
		int move;
		do {	
			printf("Make a move:");
			scanf("%i", &move);
		}
		while(isValid(&board,move-1) == 0);
		makeMove(&board, move-1, OPPONENT);
		struct evaluation e = minimax(&board, AI, 7, -INFINITY, INFINITY);
		makeMove(&board, e.bestMove, AI);
		printf("Best move is %i with a score of %i\n",e.bestMove+1,e.eval);	
		printBoard(&board);	
		if (isTerminal(&board) == 1) {
			printf("Game Over!\n");
			gameOver = 1;
		}
	}
}

int isValid(int (*board)[6][7], int column){
	if (column > 6) {
		return 0;
	} else {
		for(int i=0;i<6;i++){
			if ((*board)[i][column] == 0) {
				return 1;
			}
		}
	}
	return 0;
}

void makeMove(int (*board)[6][7], int column, int player){
	for(int i=0;i<6;i++){
		if ((*board)[i][column] == 0) {
			(*board)[i][column] = player;
			break;
		}	
	}
}

int isTerminal(int (*board)[6][7]){
	for(int y=0; y<6; y++) {
		for(int x=0; x<4; x++) {
			if (abs((*board)[y][x]+(*board)[y][x+1]+(*board)[y][x+2]+(*board)[y][x+3]) == 4){
				//printf("horizontal");
				return 1;
			}
		}
	}
	
	for(int x=0; x<7; x++) {
		for(int y=0; y<3; y++) {
			if(abs((*board)[y][x]+(*board)[y+1][x]+(*board)[y+2][x]+(*board)[y+3][x]) == 4){
				//printf("vertical");
				return 1;
			}
		}
	}
	
	for(int x=0; x<4; x++) {
		for(int y=0; y<3; y++) {
			if(abs((*board)[y][x] + (*board)[y+1][x+1] + (*board)[y+2][x+2] + (*board)[y+3][x+3]) == 4 || abs((*board)[5-y][x] + (*board)[5-y-1][x+1] + (*board)[5-y-2][x+2] + (*board)[5-y-3][x+3]) == 4){
				//printf("%i\n", abs((*board)[y][x] + (*board)[y+1][x+1] + (*board)[y+2][x+2] + (*board)[y+2][x+2]));
				//printf("%i\n", abs((*board)[5-y][x] + (*board)[5-y-1][x+1] + (*board)[5-y-2][x+2] + (*board)[5-y-3][x+3]));
				
				return 1;
			}
		}
	}
	
	return 0;
}

int evaluateSegment(int s1, int s2, int s3, int s4){
	int count[] = {0,0,0};
	int segment[] = {s1,s2,s3,s4};
	
	for(int i = 0; i < 4; i++){
		switch(segment[i]){
			case -1:
				count[0]++;
				break;
			case 0:
				count[1]++;
				break;
			case 1:
				count[2]++;
				break;
		}
	}
	
	
	if(count[0] == 4){
		//printf("lose\n");
		return -INFINITY;
	}
	if(count[2] == 4){
		//printf("win %i\n", INFINITY);
		return INFINITY;
	}
	
	if(count[0] == 3 && count[1] == 1){
		//printf("Segment: 3\n");
		return -5;
	}
	if(count[2] == 3 && count[1] == 1){
		//printf("Segment: 3\n");
		return 5;
	}
	
	if(count[0] == 2 && count[1] == 2){
		//printf("Segment: 2\n");
		return -2;
	}
	if(count[2] == 2 && count[1] == 2){
		//printf("Segment: 2\n");
		return 2;
	}
	return 0;
}

int evaluatePosition(int (*board)[6][7]){
	int scoreAdd = 0;
	int score = 0;
	for(int y=0; y<6; y++) {
		for(int x=0; x<4; x++) {
			scoreAdd = evaluateSegment((*board)[y][x],(*board)[y][x+1],(*board)[y][x+2],(*board)[y][x+3]);
			if (scoreAdd == (int)INFINITY || scoreAdd == (int)-INFINITY){
				return scoreAdd;
			} else {
				score += scoreAdd;
			}
			//printf("score: %i\n", score);
		}
	}
	
	for(int x=0; x<7; x++) {
		for(int y=0; y<3; y++) {
			scoreAdd = evaluateSegment((*board)[y][x],(*board)[y+1][x],(*board)[y+2][x],(*board)[y+3][x]);
				//printf("score: %i\n", score);
			if (scoreAdd == (int)INFINITY || scoreAdd == (int)-INFINITY){
				return scoreAdd;
			} else {
				score += scoreAdd;
			}
		}
	}
	
	for(int x=0; x<4; x++) {
		for(int y=0; y<3; y++) {
			scoreAdd = evaluateSegment((*board)[y][x],(*board)[y+1][x+1],(*board)[y+2][x+2],(*board)[y+3][x+3]);
			if (scoreAdd == (int)INFINITY || scoreAdd == (int)-INFINITY){
				return scoreAdd;
			} else {
				score += scoreAdd;
			}	
						
			//printf("score: %i\n", score);
			
			scoreAdd = evaluateSegment((*board)[5-y][x],(*board)[5-y-1][x+1],(*board)[5-y-2][x+2],(*board)[5-y-3][x+3]);
			if (scoreAdd == (int)INFINITY || scoreAdd == (int)-INFINITY){
				return scoreAdd;
			} else {
				score += scoreAdd;
			}
		}
	}
	
	return score;
}

struct evaluation minimax(int (*board)[6][7], int player, int depth, int alpha, int beta){
	int value;
	int bestMove;
	
	if(depth == 0 || isTerminal(board) == 1) {
		struct evaluation e = {evaluatePosition(board), 0};
		return e;
	}
	
	if (player == 1) {
		value = -INFINITY;
		bestMove = 0;
		
		for(int i = 0; i < 7; i++) {
			//usleep(500000);
			if(isValid(board, i) == 1){
				if (depth == 7){
					printf("---Column %i---\n", i+1);
				}
				int boardCopy[6][7];
				memcpy(boardCopy, board, boardSize);
				makeMove(&boardCopy, i, player);
				struct evaluation e = minimax(&boardCopy, player*-1, depth-1, alpha, beta);
				if (depth==7){
				printf("%i\n", e.eval);
				printf("%i\n", e.bestMove);
				//printf("%i\n", value);
				//printf("%i\n", e.eval > value);
			}
				//alpha = fmaxf(value, alpha);
				
				//if(alpha >= beta){
				//	break;
				//}
				
				if(e.eval > value){
					//if (depth ==7){
					//	printf("Eval: %i beat %i\n", e.eval, value);
					//	printf("Best move: %i beat %i\n", i, bestMove);
					//	printf("------------\n");
					//}
					value = e.eval;
					bestMove = i;
				
				}	
			}
		}
		//printf("Best move: %i\n", bestMove);
	}
	
	if (player == -1) {
		value = INFINITY;
		bestMove = 0;
		
		for(int i = 0; i < 7; i++) {
			
			if(isValid(board, i) == 1){
				int boardCopy[6][7];
				memcpy(boardCopy, board, boardSize);
				makeMove(&boardCopy, i, player);
				struct evaluation e = minimax(&boardCopy, player*-1, depth-1, alpha, beta);
				//beta = fminf(value, alpha);
				
				//if(alpha >= beta){
				//	break;
				//}
				
				if(e.eval < value){
					value = e.eval;
					bestMove = i;
					//printf(Best move is ....)
				}
			}
		}
	}
				//printf("%i\n", value);
				//printf("%i\n", bestMove);	
	struct evaluation e = {value, bestMove};
	return e;
}

void printBoard(int (*board)[6][7]){
	printf("----Board----\n");
	for(int i=0;i<6;i++){
		for(int x=0;x<7;x++){
			if ((*board)[5-i][x] == 1){
				printf("\033[93m X");
			}
			if ((*board)[5-i][x] == -1){
				printf("\033[31m O");
			} 
			if ((*board)[5-i][x] == 0){
				printf("\033[97m  ");
			}
		}
		printf("\033[97m\n");
		//printf("%i %i %i %i %i %i %i\n", (*board)[i][0],(*board)[i][1],(*board)[i][2],(*board)[i][3],(*board)[i][4],(*board)[i][5],(*board)[i][6]);
	}
	printf("\033[97m -------------\n");
	printf(" 1 2 3 4 5 6 7\n");
	printf("\033[0m");
}

File: connect4_ai.py
import copy 

OPPONENT = -1
AI = 1

FIRST_PLAYER = None
SECOND_PLAYER = None

board = [[0, 1, 1,-1, 1,-1, 0],
	     [0, 1, 0, 1,-1, 0, 0],
	     [0,-1, 0,-1, 1, 0, 0],
		 [0,-1, 0,-1,-1, 0, 0],
	     [0, 1, 0, 1, 1, 0, 0],
	     [0, 0 ,0, 0,-1, 0, 0]]
         
boards = [
		 [[0,0,0,0,0,0,0],
           [0,0,0,0,0,0,0],
           [0,0,0,0,0,0,0],
           [0,0,0,0,0,0,0],
           [0,0,0,0,0,0,0],
           [0,0,0,0,0,0,0]]
         ]

f = open("hashmaps_depth7.csv")
hashmaps = [[float(n) for n in i.split(",")] for i in f.read().split("\n")[0:-1]]

def valid_move(board, column):
	if column > 6:
		return False
	for i in board:
		if i[column] == 0:
			return True
	return False

def whose_turn(board):
	counter = 0
	for y in board:
		for x in y:
			if x == 0:
				counter += 1
	if counter % 2 == 0:
		return FIRST_PLAYER
	else: 
		return SECOND_PLAYER
		
def input_board(board,column, player):
	for i in board:
		if i[column] == 0:
			i[column] = player
			break
	return board

def to_int(board):
	ai_board = copy.deepcopy(board)
	opponent_board = copy.deepcopy(board)
	ai_int = ""
	opponent_int = ""
	for y in ai_board:
		for x in y:
			if x == -1:
				x = 0
			ai_int += str(x)
				
	for y in opponent_board:
		for x in y:
			if x == 1:
				x = 0
			if x == -1:
				x = 1
			opponent_int += str(x)			
	return int(ai_int, 2), int(opponent_int, 2)

def terminal(board):
	#horizontal
	for y in range(6):
		for x in range(4):
			if abs(sum(board[y][x:x+4])) == 4:
				return True
				
	#vertical
	for x in range(7):
		for y in range(3):
			if abs(sum([i[x] for i in board[y:y+4]])) == 4:
				return True	
	#diagonals
	for x in range(4):
		for y in range(3):
			if abs(sum(board[y+i][x+i] for i in range(4))) == 4 or abs(sum(board[::-1][y+i][x+i] for i in range(4))) == 4:
				return True
	counter = 0
	for y in board:
		for x in y:
			if x != 0:
				counter += 1
	
	if counter == 42:
		return True
	else:
		return False
		
def evaluate_segment(segment):
	global AI
	global OPPONENT
	score = 0
	if segment.count(AI) == 4:
		score = float("inf")
	if segment.count(AI) == 3 and segment.count(0) == 1:
		score += 5
	if segment.count(AI) == 2  and segment.count(0) == 2:
		score += 2
	if segment.count(OPPONENT) == 4:
		score = -float("inf")
	if segment.count(OPPONENT) == 3  and segment.count(0) == 1:
		score -= 5
	if segment.count(OPPONENT) == 2  and segment.count(0) == 2:
		score -= 2
	return score
	
def evaluate_position(board):
	score = 0
	
	for y in range(6):
		for x in range(4):
			score += evaluate_segment(board[y][x:x+4])
	for x in range(7):
		for y in range(3):
			score += evaluate_segment([i[x] for i in board[y:y+4]])
	for x in range(4):
		for y in range(3):
			score += evaluate_segment([board[y+i][x+i] for i in range(4)])
			score += evaluate_segment([board[::-1][y+i][x+i] for i in range(4)])
	return score

def minimax(board, player, depth, alpha, beta):
	#print(board)
	'''for y in range(6):
		for x in range(7):
			print("{0:+}".format(board[y][x]), end="")
		print("")
	print("-----" +str(depth)+"-------")'''
	global AI
	global OPPONENT
	global hashmaps
	
	'''if depth == 7:
		for i in hashmaps:
			if to_int(board) == tuple(i[0:2]):
				return i[3], int(i[2])'''
	
	if depth == 0 or terminal(board):
		return evaluate_position(board), None
	
	if player == AI:
		value = -float("inf")
		best_move = 0
		for column in range(0,7):
			if valid_move(board, column):
				board_copy = copy.deepcopy(board)
				board_copy = input_board(board_copy,column,AI)
				new_value,move = minimax(board_copy, OPPONENT, depth - 1, alpha, beta)
				alpha = max(value, alpha)
				if alpha >= beta:
					break
				if new_value > value:
					value = new_value
					best_move = column
					print("Best Move is :" + str(best_move+1) + "---->" + "{0:+}".format(value) if depth == 7 else "", end="\r")
		return value, best_move
		
	if player == OPPONENT:
		value = float("inf")
		best_move = 0
		for column in range(0,7):
			if valid_move(board, column):
				board_copy = copy.deepcopy(board)
				board_copy = input_board(board_copy,column,OPPONENT)
				new_value,move = minimax(board_copy, AI, depth - 1, alpha, beta)
				beta = min(value, beta)
				if alpha >= beta:
					break
				if new_value < value:
					value = new_value
					best_move = column
		return value, best_move
#Gameplay

def main():
	global OPPONENT
	global AI
	global board
	print(minimax(board, AI, 5, -float("inf"), float("inf")))
	'''while True:
		turn = input("Who Goes first?:")
		if turn == "me":
			FIRST_PLAYER = OPPONENT
			SECOND_PLAYER = AI
			break
		if turn == "you":
			FIRST_PLAYER = AI
			SECOND_PLAYER = OPPONENT
			break
	if turn == "me":
		while True:
			move = int(input("Make a move:"))
			if valid_move(board,move-1):
				board = input_board(board,move-1,OPPONENT)
				break
	
	score_log = []
	while True:
		value, ai_move = minimax(board, AI, 7, -float("inf"), float("inf"))
		print("")
		print("My move is: " + str(ai_move+1))
		print("The score predicted is: " + str(value))
		score_log.append(value)
		#print(to_int(board))
		board = input_board(board, ai_move, AI)
		for y in range(6):
			for x in range(7):
				print(("\033[93m X" if board[::-1][y][x] == 1 else "\033[31m O" if board[::-1][y][x] == -1 else "\033[97m  "),end="" )#+ "{0:+}".format(board[::-1][y][x]), end="")
			print("")
		print("\033[97m -------------")
		print(" 1 2 3 4 5 6 7")
		print("\033[0m")
		if terminal(board):
			print("Game over!")
			print("Score Log:", end="")
			for x in score_log:
				print(str(x) + " ", end="")
			break
		while True:
			move = int(input("Make a move:"))
			if valid_move(board, move-1):
				board = input_board(board, move-1, OPPONENT)
				break
			print("Invalid Move! Try Again")
		if terminal(board):
			print("Game over!")
			break'''


main()

def train(board):
	f = open("hashmaps_depth7(1).csv", "a")
	value, ai_move = minimax(board, AI, 7, -float("inf"), float("inf"))
	f.write(str(to_int(board)[0]) + "," + str(to_int(board)[1]) + "," + str(ai_move) + "," + str(value) + "\n")
	print(f"Best move is {ai_move+1}")
	return ai_move


def pt(board):
	for y in range(6):
		for x in range(7):
			print(("\033[93m X" if board[::-1][y][x] == 1 else "\033[31m O" if board[::-1][y][x] == -1 else "\033[97m  "),end="" )#+ "{0:+}".format(board[::-1][y][x]), end="")
		print("")
	print("\033[97m -------------")
	print(" 1 2 3 4 5 6 7")
	print("\033[0m")
 
def ok(t):
	t1= t[0]
	t2= t[1]
	b1 = str(bin(t1))[2::].zfill(42)
	b2 = str(bin(t2))[2::].zfill(42)
	import numpy as np
	b1 = [int(x) for x in b1]
	b2 = [int(x) for x in b2]
	b=[]
	for i in range(0,42):
		if b1[i] == 1:
			b.append(1)
		elif b2[i] == 1:
			b.append(-1)
		else:
			b.append(0)
	pt(np.array(b).reshape(6,7))

'''import time
ok(tuple(int(i) for i in hashmaps[5][0:2]))
for x in range(7):
	print("At Depth: " + str(x+1))
	c = 1
	boards_copy = copy.deepcopy(boards)
	for b in range(len(boards_copy)):
		pt(boards[b])
	
	for b in range(len(boards_copy)):
		if terminal(boards_copy[b]) == False:
			for i in range(7):
				if valid_move(boards_copy[b],i):
					pt(boards[b])
					boards.append(input_board(copy.deepcopy(boards_copy[b]),i,-1))
					pt(boards[-1])
					print(f"Dropped on {i+1}")
		boards.remove(boards[b])

	for b in range(len(boards)):
		print("Depth " + str(x+1) + " and Board " + str(c))
		pt(boards[b])
		move = 1#train(b)
		boards[b] = input_board(copy.deepcopy(boards[b]),move,1)
		pt(boards[b])
		c += 1
		
		time.sleep(0.1)
'''
